base/compiler/typeinfer.jl

```diff
@@ -1001,48 +1014,82 @@ function typeinf_frame(interp::AbstractInterpreter, mi::MethodInstance, run_opti
     return frame
 end
 
+# We only care about types/effects, no source required
+const SOURCE_MODE_NOT_REQUIRED = 0x0
+
+# We need something that can be invoked or compiled (i.e. constabi or inferred)
+const SOURCE_MODE_ABI = 0x1
+
+# We need source, even for constabi
+const SOURCE_MODE_FORCE_SOURCE = 0x2
+
+function ci_has_source(code::CodeInstance)
+    inf = @atomic :monotonic code.inferred
+    return isa(inf, CodeInfo) || isa(inf, String)
+end
+
+"""
+    ci_has_abi(code::CodeInstance)
+
+Determine whether this CodeInstance is something that could be invoked if we gave it
+to the runtime system (either because it already has an ->invoke ptr, or because it
+has source that could be compiled).
+"""
+function ci_has_abi(code::CodeInstance)
+    ci_has_source(code) && return true
```

Comment:
This constraint should be documented, as right now there is nothing that requires the system to preserve this source (in particular, there already exists a long gap in time between when the compiler runs over this, and when the jit might decide to set the invoke pointer, if ever).

Possibly a way to ensure this predicate is accurate is to add another atomic field, so that `ci_has_abi` becomes rather a `set_ci_has_abi!` mutation predicate, which sets a field to require that the `inferred` pointer must be preserved during the current runtime session (of course, serialization is permitted to clear the inferred and source fields, since it must be permitted to, so it also may clear the `set_ci_has_abi` field also

Reply:
How about before deleting the code, we set `invoke` to something that tries to acquire the codegen lock and once the codegen lock has been acquired re-invokes itself. Aren't we guaranteed that by the end of the current codegen, the invoke ptr will be set?

Reply:
We currently won't set the invoke pointer until the JIT compiles it, which is necessarily some arbitrary amount time in the future after codegen finished with it. Trying to put some jl_call_trampoline-like thing here seems like it may work (we have used that design in the past), though it seemed to have some complicated threading and serialization implications last time we had that.